include/nbl/builtin/hlsl/workgroup/arithmetic.hlsl

```diff
@@ -0,0 +1,128 @@
+    T operator()(T value)
+    {
+		SharedAccessor accessor;
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(true, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+        wsh(value);
+        WorkgroupScanTail<T, Binop, SharedAccessor> wst = WorkgroupScanTail<T, Binop, SharedAccessor>::create(false, Binop::identity(), wsh.firstLevelScan, wsh.lastInvocation, wsh.scanStoreIndex);
+		T retVal = wst();
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+		return retVal;
+    }
+};
+
+template<typename T, class Binop, class SharedAccessor>
+struct exclusive_scan
+{
+    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {}; // Yes, inclusive scan subgroup op is used for exclusive workgroup ops
+    T operator()(T value)
+    {
+		SharedAccessor accessor;
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(true, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+        wsh(value);
+        WorkgroupScanTail<T, Binop, SharedAccessor> wst = WorkgroupScanTail<T, Binop, SharedAccessor>::create(true, Binop::identity(), wsh.firstLevelScan, wsh.lastInvocation, wsh.scanStoreIndex);
+		T retVal = wst();
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+		return retVal;
+    }
+};
+
+#define WSHT WorkgroupScanHead<uint, subgroup::inclusive_scan<uint, binops::add<uint> >, SharedAccessor>
+#define WSTT WorkgroupScanTail<uint, binops::add<uint>, SharedAccessor>
+template<class SharedAccessor>
+uint ballotScanBitCount(in bool exclusive)
+{
+	SharedAccessor accessor;
+	const uint _dword = getDWORD(gl_LocalInvocationIndex);
+	const uint localBitfield = accessor.main.get(_dword);
+	uint globalCount;
+	{
+		uint localBitfieldBackup;
+		if(gl_LocalInvocationIndex < uballotBitfieldCount)
+		{
+			localBitfieldBackup = accessor.main.get(gl_LocalInvocationIndex);
+		}
+		// scan hierarchically, invocations with `gl_LocalInvocationIndex >= uballotBitfieldCount` will have garbage here
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+		
+		WSHT wsh = WSHT::create(true, 0u, uballotBitfieldCount);
+		wsh(countbits(localBitfieldBackup));
+		
+		WSTT wst = WSTT::create(true, 0u, wsh.firstLevelScan, wsh.lastInvocation, wsh.scanStoreIndex);
+		wst();
+		
+		// fix it (abuse the fact memory is left over)
+		globalCount = _dword != 0u ? accessor.main.get(_dword) : 0u;
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+		
+		// restore because the counting process has changed the ballots in the shared mem
+		// and we might want to use them further
+		if(gl_LocalInvocationIndex < uballotBitfieldCount)
+		{
+			accessor.main.set(gl_LocalInvocationIndex, localBitfieldBackup);
+		}
+		accessor.main.workgroupExecutionAndMemoryBarrier();
```

Comment:
you can stop the `localBitfieldBackup` shenanigans if you allow "stateful" Accessors

the reason I had this was because I had no way to cleanly customize what gets used as scratch memory so all scratch memory was aliased

Reply:
still outstanding

Reply:
still outstanding